File: solutions/Virtual-Assistant/src/csharp/Microsoft.Bot.Solutions/Middleware/EventDebuggerMiddleware.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Microsoft.Bot.Solutions.Middleware
{
    public class EventDebuggerMiddleware : IMiddleware
    {
        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
        {
            var activity = turnContext.Activity;

            if (activity.Type == ActivityTypes.Message)
            {
                var text = activity.Text;
                var value = activity.Value?.ToString();

                if (!string.IsNullOrEmpty(text) && text.StartsWith("/event:"))
                {
                    var json = text.Split("/event:")[1];
                    var body = JsonConvert.DeserializeObject<Activity>(json);

                    turnContext.Activity.Type = ActivityTypes.Event;
                    turnContext.Activity.Name = body.Name;
                    turnContext.Activity.Value = body.Value;
                }

                if (!string.IsNullOrEmpty(value) && value.Contains("event"))
                {
                    var root = JObject.Parse(value);
                    var eventActivity = (JObject)root["event"];
                    turnContext.Activity.Type = ActivityTypes.Event;

                    if (eventActivity["name"] != null)
                    {
                        turnContext.Activity.Name = eventActivity["name"].ToString();
                    }

                    if (eventActivity["text"] != null)
                    {
                        turnContext.Activity.Text = eventActivity["text"].ToString();
                    }

                    if (eventActivity["value"] != null)
                    {
                        turnContext.Activity.Value = eventActivity["value"].ToString();
                    }
                }
            }

            await next(cancellationToken).ConfigureAwait(false);
        }
    }
}

File: sdk/typescript/libraries/bot-solutions/src/middleware/eventDebuggerMiddleware.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { Middleware, TurnContext } from 'botbuilder';
import { Activity, ActivityTypes } from 'botframework-schema';

export class EventDebuggerMiddleware implements Middleware {
    public async onTurn(turnContext: TurnContext, next: () => Promise<void>): Promise<void> {
        const activity: Activity = turnContext.activity;

        if (activity.type === ActivityTypes.Message && activity.type !== undefined && activity.type.trim().length > 0) {
            const text: string = activity.text;
            const value: string = JSON.stringify(activity.value);

            if (text !== undefined && text.trim().length > 0 && text.startsWith('/event:')) {
                const json: string = text.substr('/event:'.length);
                const body: Activity = JSON.parse(json);

                turnContext.activity.type = ActivityTypes.Event;

                turnContext.activity.name = body.name || turnContext.activity.name;
                turnContext.activity.text = body.text || turnContext.activity.text;
                turnContext.activity.value = body.value || turnContext.activity.value;
            }

            if (value !== undefined && value.trim().length > 0 && value.includes('event')) {
                const body: { event: { name: string; text: string; value: string }} = JSON.parse(value);

                turnContext.activity.type = ActivityTypes.Event;
                if (body.event !== undefined) {
                    if (body.event.name !== undefined) {
                        turnContext.activity.name = body.event.name;
                    }

                    if (body.event.text !== undefined) {
                        turnContext.activity.text = body.event.text;
                    }

                    if (body.event.value !== undefined) {
                        turnContext.activity.value = body.event.value;
                    }
                }
            }
        }

        await next();
    }
}
